[1.7] Main menu improvements

GeorgeK1ng opened the pull request:
1) Added option to use custom logo and sublogo on Main menu, Scenario selection and Loading screen
2) Added option to use custom loading bar frame on Loading screen

IvanSavenko commented on config/mainmenu.json:
```diff
@@ -1,22 +1,60 @@
+	{
+		"background" : ["gamselb0", "gamselb1"],
+		
+		//"logo": {
+		//	"name": ["logo/logo0", "logo/logo1", "logo/logo2"], 
+		//	"x": 20, "y": 20
+		//},
+		//"sublogo": {
+		//	"name": ["sublogo/sublogo0", "sublogo/sublogo1", "sublogo/sublogo2"], 
+		//	"x": 20, "y": 120
+		//}
```
Perhaps format like this would be more suiltable? Why VCMI should care about what these images represent? It could be logo or could be something else:
```json
    "images": [
        {
            "x": 20, 
            "y": 20,
            "name": ["logo/logo0", "logo/logo1", "logo/logo2"] // randomly selected
        },
        {
            "x": 20, 
            "y": 120,
            "name": ["sublogo/sublogo0", "sublogo/sublogo1", "sublogo/sublogo2"] // randomly selected
        }
```

GeorgeK1ng replied:
Looks better, but I have no idea how to do that in code as their variables are directly in headers.

GeorgeK1ng replied:
Something like this is my goal. Logo = Main Heroes 3 text and SubLogo expansion like text.
![image](https://github.com/user-attachments/assets/6e59508a-7453-4152-8396-e10ef4cefa47)
![image](https://github.com/user-attachments/assets/bd561221-29e4-4d97-9b50-870d5a155678)
![image](https://github.com/user-attachments/assets/c0b0497c-4037-4129-8189-16abeac28759)

IvanSavenko commented on client/lobby/CSelectionBase.cpp:
```diff
@@ -93,13 +93,27 @@ CSelectionBase::CSelectionBase(ESelectionScreen type)
 	{
-		const JsonVector & bgNames = CMainMenuConfig::get().getConfig()["game-select"].Vector();
+		const JsonNode& gameSelectConfig = CMainMenuConfig::get().getConfig()["scenario-selection"];
+		const JsonVector& bgNames = gameSelectConfig["background"].Vector();
```
Please add fallback so game won't crash if player has 1.6 mod

GeorgeK1ng replied:
Done in latest commit

IvanSavenko commented on client/mainmenu/CMainMenu.cpp:
```diff
@@ -88,6 +88,18 @@ CMenuScreen::CMenuScreen(const JsonNode & configNode)
 	pos = background->center();
 
+	const auto& logoConfig = config["logo"];
+	if (!logoConfig.isNull() && logoConfig["name"].isVector() && !logoConfig["name"].Vector().empty()) {
```
- place opening bracket on new line
- there is no need to check json node type. Accessing json using wrong type is well-defined behavior and will return default-constructed value, in this case - empty vector
```suggestion
	if (!logoConfig["name"].Vector().empty()) 
	{
```

GeorgeK1ng replied:
Simplified code in latest commit

IvanSavenko commented on client/mainmenu/CMainMenu.cpp:
```diff
@@ -88,6 +88,18 @@ CMenuScreen::CMenuScreen(const JsonNode & configNode)
+	}
+
+	const auto& sublogoConfig = config["sublogo"];
+	if (!sublogoConfig.isNull() && sublogoConfig["name"].isVector() && !sublogoConfig["name"].Vector().empty()) {
```
Same here

GeorgeK1ng replied:
Simplified code in latest commit

IvanSavenko commented on client/mainmenu/CMainMenu.cpp:
```diff
@@ -729,25 +732,51 @@ CLoadingScreen::CLoadingScreen(ImagePath background)
+	// Load background
+	const auto& backgroundConfig = conf["background"];
+	if (backgroundConfig.isVector() && !backgroundConfig.Vector().empty())
+		this->background = std::make_shared<CPicture>(ImagePath::fromJson(*RandomGeneratorUtil::nextItem(backgroundConfig.Vector(), CRandomGenerator::getDefault())));
```
No need to explicitly write `this` unless necessary
```suggestion
		background = std::make_shared<CPicture>(ImagePath::fromJson(*RandomGeneratorUtil::nextItem(backgroundConfig.Vector(), CRandomGenerator::getDefault())));
```

GeorgeK1ng replied:
Can't be compiled without **this->**

GeorgeK1ng commented:
Will review all mentioned things. Seems I need to make configurable more hardcoded images to make full Demo support for main menu aswell.

GeorgeK1ng commented:
I have solved all mentioned things now. Now are needed images stored in "images" block instead, simplified code, added required 1.6 fallback. Removed needs of usage of **this->**. Now sounds for me ready to merge.
